### PR TITLE
Revised support for AO/ENCODING in CREATE TABLE .. LIKE INCLUDING

### PR DESCRIPTION
**Pull request by l-wang**
Previously, commit 951065c48cc1e636ca9bf54bde440c93482ad9bc, as part of
PR #7609 added support for the encodings of append-optimized table by
overloading the existing upstream option INCLUDING STORAGE, which in
upstream is used only for toast management. This Overloading let to
limitations and inconsistencies, especially when access method and
reloptions are explicitly specified, or when there are multiple CRATE
TABLE LIKE INCLUDING STORAGE clauses. Commit
a053cc5c1f90f300c194b5986b02f63d3fb80569 records some of the
limitations.

**This commit reimplemented support for AO encodings by introducing three
new keywords as INCLUDING options:**

**- INCLUDING ENCODING: includes per-attributed encoding for ao_column
  tables**

**- INCLUDING RELOPT: includes general reloptions,** such as "blocksize",
  "compresslevel", "compresstype" for a ao_row or ao_column tables,
  "fillfactor" for heap tables, as well as more general reloptions

**- INCLUDING AM: includes access method**

NB:
- _INCLUDING AM can't be specified if access method is already explicitly
  specified in the same CREATE TABLE statement by USING or WITH clause_

- _INCLUDING RELOPT can't be specified if reloptions are already explicitly
  specified in the same CREATE TABLE statement by WITH clause_

- _If multiple INCLUDING AM clauses or multiple INCLUDING RELOPT clauses
  are specified, take the last one._

Discussion: https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/mOPeBxFm43w

Reviewers: Does the new syntax make sense?

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/parser/parse_utilcmd.c [hunk 1]
```diff
@@ -1232,21 +1238,30 @@ transformTableLikeClause(CreateStmtContext *cxt, TableLikeClause *table_like_cla
...
+	{
+		MemoryContext oldcontext;
+		/*
+		 * As we are modifying the utility statement we must make sure these
+		 * DefElem allocations can survive outside of this context.
+		 */
+		oldcontext = MemoryContextSwitchTo(CurTransactionContext);
+
+		// TODO: allow multiple LIKE clauses and take last specification
+		if (stmt->accessMethod && strcmp(stmt->accessMethod, get_am_name(relation->rd_rel->relam)) != 0)
```
Comment (huansong) [inline src/backend/parser/parse_utilcmd.c:1252]: lazy asking: is `with(appendonly=...,orientation=...)` going to be rejected here too (which they should)?

File: src/test/regress/expected/create_table_like_gp.out [hunk 2]
```diff
@@ -7,11 +7,23 @@ NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as
...
-CREATE TABLE t_ao_b (LIKE t_ao INCLUDING STORAGE);
+CREATE TABLE t_ao_b (LIKE t_ao INCLUDING STORAGE); -- Should create a heap table
+NOTICE:  table doesn't have 'DISTRIBUTED BY' clause, defaulting to distribution columns from LIKE table
+CREATE TABLE t_ao_b1 (LIKE t_ao INCLUDING ENCODING); -- Should error
+NOTICE:  table doesn't have 'DISTRIBUTED BY' clause, defaulting to distribution columns from LIKE table
+ERROR:  ENCODING clause only supported with column oriented tables
+CREATE TABLE t_ao_b2 (LIKE t_ao INCLUDING RELOPT); -- Should error
+NOTICE:  table doesn't have 'DISTRIBUTED BY' clause, defaulting to distribution columns from LIKE table
+ERROR:  unrecognized parameter "blocksize"
+CREATE TABLE t_ao_b3 (LIKE t_ao INCLUDING RELOPT INCLUDING AM); -- Should create aocs table with copied reloptions
```
Comment (huansong) [inline src/test/regress/expected/create_table_like_gp.out:18]: Let's also add a case for `INCLUDING ALL` for ao_column table (we have one for ao_row) which should copy over all of AM, encoding and relopt.

**Review by huansong — COMMENTED** [review]
The usage LGTM. Thanks for driving the efforts to make them more clear!

File: src/backend/parser/parse_utilcmd.c (hunk 1 above)
Comment (l-wang) [reply src/backend/parser/parse_utilcmd.c:1252]: Yes! At this point, parser function `greenplumLegacyAOoptions()` has already taken `appendonly=...,orientation=...` out of the `stmt->options` list and set `stmt->accessmethod` accordingly.

File: src/test/regress/expected/create_table_like_gp.out [hunk 3]
```diff
@@ -20,16 +32,19 @@ NOTICE:  table doesn't have 'DISTRIBUTED BY' clause, defaulting to distribution
 RESET gp_default_storage_options;
 -- Verify created tables and attributes
 select relname, reloptions from pg_class where relname LIKE 't_ao%' order by relname;
-  relname   |                            reloptions                             
-------------+-------------------------------------------------------------------
- t_ao       | {blocksize=32768,compresslevel=0,compresstype=none,checksum=true}
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/test/regress/expected/create_table_like_gp.out:25]: Good call!

**Comment by jimmyyih** [comment]
Briefly looking at the new syntax and the test SQL examples (didn't look too closely at the actual logic change), the change makes sense to me.

File: src/backend/parser/parse_utilcmd.c (hunk 1 above)
Comment (l-wang) [reply src/backend/parser/parse_utilcmd.c:1245]: Edit: I added check in the parser to help reject cases like `USING <am>` and `with(appendonly=...,orientation=...)` earlier. So by the time we get here, if `stmt->accessMethod` is set, it has to come from a previous INCLUDING AM clause.

File: src/backend/parser/gram.y [hunk 4]
```diff
@@ -19861,6 +19879,52 @@ check_expressions_in_partition_key(PartitionSpec *spec, core_yyscan_t yyscanner)
...
+	{
+		Node	*element = lfirst(elements);
+		if (nodeTag(element) == T_TableLikeClause)
+		{
+			if (((TableLikeClause *) element)->options & CREATE_TABLE_LIKE_RELOPT)
+				return true;
+		}
+	}
+	return false;
+}
```
Comment (huansong) [inline src/backend/parser/gram.y:19926]: nit: can we do one pass for both? One idea is to have a:
```
greenplumCreateTableRejectLike(const List *tableElts, bool reject_am, bool reject_relopt)
```
which rejects `AM`/`RELOPT` depending on the parameters (and prints `ERROR` when found). And the caller simply does:
```
greenplumCreateTableRejectLike(n->tableElts, n->accessMethod != NULL, n->options != NULL);
```

File: src/backend/parser/parse_utilcmd.c [hunk 5]
```diff
@@ -1232,65 +1238,28 @@ transformTableLikeClause(CreateStmtContext *cxt, TableLikeClause *table_like_cla
...
-		MemoryContextSwitchTo(oldcontext);
+	/* Copy reloptions if requested */
+	if (table_like_clause->options & CREATE_TABLE_LIKE_RELOPT)
+	{
+		/* Allow multiple LIKE INCLUDING RELOPT clauses and take last specification */
+		if (stmt->options)
+			ereport(WARNING,
+					(errcode(ERRCODE_DUPLICATE_OBJECT),
+						errmsg("Multiple INCLUDING RELOPT is specified, applying the last specification only LIKE %s",
+							   RelationGetRelationName(relation))));
```
Comment (huansong) [inline src/backend/parser/parse_utilcmd.c:1261]: question though: do we really need the WARNING? Are we throwing it for any existing cases?

**Review by huansong — COMMENTED** [review]
Mostly LGTM, the test coverage especially for corner cases is really nice.

Just wondering about one thing: do we really need to throw `WARNING` for multiple `INCLUDING AM|RELOPT`: it seems we don't do the same for any existing cases (e.g. `COMMENTS`, `CONSTRAINTS`, etc..) right?

In fact, it seems that the existing cases might not have the same problem: they are all column-level properties. So since we don't allow multiple `LIKE` tables to have duplicated column, we don't need to worry about multiple specifications of the same property for the same column.

That brings me to wonder does this statement in [upstream doc](https://www.postgresql.org/docs/12/sql-createtable.html):
```
If multiple specifications are made for the same kind of object, the last one is used
```
means multiple `LIKE`, or multiple `INCLUDING|EXCLUDING` in the same `LIKE`?

In this PR we are applying the above rule to multiple `LIKE`, but should we even apply the rule if it only means for multiple `INCLUDING|EXCLUDING` in the same `LIKE`? E.g.
```sql
postgres=# create table t1(a int, b int, check (b>0));
CREATE TABLE
postgres=# \d+ t1
                                    Table "public.t1"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description
--------+---------+-----------+----------+---------+---------+--------------+-------------
 a      | integer |           |          |         | plain   |              |
 b      | integer |           |          |         | plain   |              |
Check constraints:
    "t1_b_check" CHECK (b > 0)
Distributed by: (a)
Access method: heap

postgres=# create table t2(like t1 including constraints excluding constraints);
CREATE TABLE
postgres=# \d+ t2
                                    Table "public.t2"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description
--------+---------+-----------+----------+---------+---------+--------------+-------------
 a      | integer |           |          |         | plain   |              |
 b      | integer |           |          |         | plain   |              |
Distributed by: (a)
Access method: heap

postgres=# create table t3(like t1 excluding constraints including constraints);
CREATE TABLE
postgres=# \d+ t3
                                    Table "public.t3"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description
--------+---------+-----------+----------+---------+---------+--------------+-------------
 a      | integer |           |          |         | plain   |              |
 b      | integer |           |          |         | plain   |              |
Check constraints:
    "t1_b_check" CHECK (b > 0)
Distributed by: (a)
Access method: heap
```
So `t3` is taking the constraints from `t1`, but `t2` is not.

So basically two questions:
1. Does the same rule apply to multiple `INCLUDING|EXCLUDING AM|RELOPT|ENCODING` in the same `LIKE`? My guess is that the existing code would already deal with that for us.
2. What should we do about multiple `LIKE ... INCLUDING AM|RELOPT|ENCODING`? I guess we can just give `ERROR`?

File: src/backend/parser/gram.y (hunk 4 above)
Comment (l-wang) [reply src/backend/parser/gram.y]: Now that we error (rather than warning) if `INCLUDING AM/REOPT` presents in multiple LIKEs, I perfer to do the error checking during transformation because:

If we do error checking in parser, we have to iterate through all of the `n->tableElts` which might not even have LIKE clauses or `INCLUDING AM/RELOPT`. This adds unnecessary overhead for very wide table with many columns and have USING/WITH clauses.

The downside of doing it during transformation is missing the error position, which I think is not a big deal as long as we have clear error message.

**Comment by l-wang** [comment]
@huansong Thank you so much for the thorough review! I addressed all of your feedbacks. I didn't do everything as you suggested but added explanations. Could you take another look and let me know?

**Review by huansong — APPROVED** [review]
A few comments need update, otherwise LGTM!

File: src/backend/parser/gram.y (hunk 4 above)
Comment (huansong) [reply src/backend/parser/gram.y]: Agreed totally.

File: src/backend/parser/parse_utilcmd.c [hunk 6]
```diff
@@ -1232,65 +1238,46 @@ transformTableLikeClause(CreateStmtContext *cxt, TableLikeClause *table_like_cla
...
+					(errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+						errmsg(
+							"LIKE %s INCLUDING RELOPT is not allowed because the reloptions of table %s is already set",
+							RelationGetRelationName(relation),
+							cxt->relation->relname)),
+					errdetail(
+						"Multiple LIKE clauses with the INCLUDING RELOPT option is not allowed.\n"
+						"Single LIKE clause with the INCLUDING RELOPT option is also not allowed if reloptions is explicitly specified by a WITH clause."),
+					errhint(
+						"Remove INCLUDING RELOPT or add EXCLUDING RELOPT from CREATE TABLE LIKE."));
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/parser/parse_utilcmd.c:1278]: EXCLUDING usage is for the case where INCLUDING ALL is specified, let me clarifily that in the error hint as well.

File: src/backend/parser/parse_utilcmd.c [hunk 7]
```diff
@@ -1232,65 +1238,46 @@ transformTableLikeClause(CreateStmtContext *cxt, TableLikeClause *table_like_cla
...
-			if (strlen(NameStr(compresstype)) > 0)
-				stmt->options = lappend(stmt->options,
-				                        makeDefElem("compresstype", (Node *) makeString(pstrdup(NameStr(compresstype))), -1));
+			Datum options;
+			options = transformRelOptions(PointerGetDatum(NULL),
+									stmt->options,
+									NULL,
+									NULL,
+									true,
+									false);
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/parser/parse_utilcmd.c]: Indeed! Thanks!
